core/tests/test_utils.py

```diff
@@ -161,6 +164,127 @@ def check_image_png_or_webp(image_string):
     return False
 
 
+class EmailMessageMock(python_utils.OBJECT):
+    """Mock for mailgun API message."""
+
+    def __init__(
+            self, sender_email, recipient_email, subject, plaintext_body,
+            html_body, bcc=None, reply_to=None):
+        self.sender = sender_email
+        self.to = recipient_email
+        self.subject = subject
+        self.body = plaintext_body
+        self.html = html_body
+        self.bcc = bcc
+        self.reply_to = reply_to
+
+
+class EmailServicesMock(python_utils.OBJECT):
+    """Mock for mailgun API"""
+
+    def __init__(self):
+        self.emails_dict = {}
+
+    def wipe_emails_dict(self):
+        """Reset email dictionary for a new test."""
+        self.emails_dict = {}
+
+    def mock_send_mail(
```

Comment:
There's lots of non-third-party-specific code here. I think what would be better to do is: limit platform/ to the actual "sending mail" action, and have the code in the email services module be as minimal as possible -- it should only contain the code that interacts with the third-party service. Compose and check everything else in the domain layer.

Also, to make things simpler for yourself, perhaps start by getting rid entirely of gae_email_services.py. You're replicating stuff from that and I think it's making life harder, since doing so is not necessary because the only thing that actually matters is the mailgun implementation.

Finally, this implementation should be self-contained and not refer to email_services at all. You are trying to swap out the one (or at most two) functions in email_services in tests, because the email service in the platforms layer should be doing nothing but making an external call, and you don't want to make external calls in the tests.

I suggest that, before doing anything else on this PR, you create a design doc that lists the functions in the final email_services module and provides the code for their implementation in full. Get sign-off on that before continuing. Then start again. Do the following:

(a) switch the default from gae_email_services to mailgun_email_services, and get rid of gae_email_services altogether. 
(b) modify mailgun_email_services to have the API and code that was approved above. (That code should handle dev/prod mode correctly.)
(c) create stubs for the functions in mailgun_email_services in this file, to be used only in tests.
(d) fix any broken tests.
(e) figure out how to generalize core/platform/models.py to allow a separate email provider implementation to be specified and to have that import the mailgun_email_services file. (It's not really reasonable to consider that a "gae" thing.)

Let me know if you have any questions about the above. Thanks!

Reply:
Hey @seanlip,
Could you clarify what this means: " You are trying to swap out the one (or at most two) functions in email_services in tests, because the email service in the platforms layer should be doing nothing but making an external call, and you don't want to make external calls in the tests."

I understand that I shouldn't be making a call to mailgun in tests but the function usage should be the same right. In tests, I should still be calling email_services.send_mail/email_services.send_bulk_mail except prior to that call, I call self.swap to swap it with my own mock function.

Reply:
@seanlip So our structure leaves send_email_to_recipients(which does the post to mailgun api) in mailgun_email_services and moves send_mail and send_bulk_mail to core/domain/email_services(). I need to mock these functions instead of the mailgun functions since most of the tests use those functions. That makes sense right? Because you mentioned that I should create stubs for functions in mailgun_email_services in point b above.

Reply:
Mock only send_email_to_recipients(). The only point of mocking is to avoid using the actual production APIs. As much of the code that's being tested should stay the same between dev/test/prod as possible, otherwise you end up testing something that has little relation to what runs in dev/prod.

Reply:
Hm, so currently all of the functions in testing uses send_mail and send_bulk_mail to send out the emails, which I moved out of mailgun_email_services into core/domain/email_services. I think it makes sense to mock these instead since this is still part of the API, just one layer up from the platform layer.

Reply:
I have already explained, in my previous comment, why that is not the way to go. You will probably need to redo all of this.